docs/Incompatibilities.md

```diff
@@ -0,0 +1,127 @@
+
+In some future version, we expect to be able to support multiple processes but there will be no isolation between them.
+This means that using multiple processes for privilege separation will not be possible.
+
+### No MMU support
+
+A conventional x86-64 system exposes multiple privilege domains ('rings') and provides a memory management unit that allows each userspace process to have a separate virtual address space.
+This includes the ability to alias pages such that the same physical page is in several different locations in the virtual address space.
+In contrast, SGX exposes a single virtual address space and does not allow page mappings to be modified.
+
+At the system call layer, this means that there are some restrictions on `mmap`:
+
+ - Mapping anonymous memory works reliably.
+ - Shared mappings are not automatically written back.
+ - `MAP_FIXED` mappings should only be done over existing mappings: the kernel and userspace share and address space.
+   It is currently possible to do `MAP_FIXED` over kernel mappings, this will be fixed in a future version.
+
+These restrictions are close to those of uCLinux and SGX-LKL will eventually use the no-MMU code from Linux.
+
+Because SGX does not grant control over the MMU, it is impossible on SGX1 hardware to securely change page permissions.
+SGX-LKL delegates this to the untrusted host.
+As a result, any `mprotect` call, or any `mmap` with permissions weaker than read-write-execute, should be considered advisory.
+These calls will work correctly *in the absence of an attacker* but a malicious host can simply ignore these calls.
+This will be addressed on hardware with Enclave Dynamic Memory Management (EDMM) support.
+
+### Cooperative threading
+
+The SGX-LKL runtime uses an enclave thread (ethread) abstraction that roughly corresponds to a virtual CPU (VCPU) in a conventional VM environment.
+In a conventional Linux system, the Linux scheduler runs userspace threads, switching to another thread either when a thread makes a blocking system call or when an interrupt fires.
+SGX; however, does not provide a mechanism for receiving timer interrupts and so cannot efficiently run a preemptive scheduler.
+
+SGX-LKL uses the lthread scheduler to provide a cooperative threading implementation, multiplexing lthreads on top of ethreads.
+The lthread scheduler running on each ethread will switch the running lthread on any system call.
+The main impact of this on userspace code is that threads that run without any system calls will not be interrupted and can consume all of the CPU resources.
+Spin locks that do not have a fallback futex path for the contended case, for example, may fail to make progress when there is a single ethread and may consume excessive amounts of CPU time in all uses.
+Threads that block on a futex will be correctly scheduled.
+
+LKL runs Linux in single-core mode and delegates scheduling of userspace threads to the lthread scheduler.
```

Comment:
What is the impact of running in "single-core mode" for userspace applications? What will `std::thread::hardware_concurrency` return? And `sysconf(_SC_NPROCESSORS_ONLN)`? Or reading `/proc/cpuinfo`? Will those return 1? Or the number of ethreads? Or lthreads? Does it work "as expected" or do users have to configure parallelisation differently in their apps for SGX-LKL (I hope not)?

Reply:
`sysconf` will return return the number of ethreads, because we override them.  Not sure about `/proc/cpuinfo`.  Only one thread can be executing a system call at a time.

Reply:
```suggestion
LKL runs Linux in single-core mode and delegates scheduling of userspace threads to the lthread scheduler.
This means that only one core can be executing system calls at any time, though userspace code can be executing on other cores at the same time.
```

Reply:
OK that makes sense now and is clear enough.